<!--

Programmer: Victoria Albanese
Filename: camera_1.launch

Purpose: Launch the Intel RealSense 435i camera
so that its pointcloud can be visualized

NOTE: this camera is mounted (from the robot's perspective)
in the front to the right, looking straight on at the object

-->

<launch>

    <!-- This launches the camera with the correct parameters -->
    <include file="$(find realsense2_camera)/launch/rs_rgbd.launch">
	<arg name="camera" value="camera_1" />
	<arg name="serial_no" value="843112074077" />
    </include>


    <!-- This transforms each resulting pointcloud into a global reference frame -->
    <node name="camera_1_tf" pkg="tf" type="static_transform_publisher"
        args="0.365 0.293 0.078 2.356 -0.025 -1.571 /world /camera_1_color_optical_frame 100"/>


    <!-- These nodelets limit the FOV of the camera (including filtering out the ground plane) -->
    <node pkg="nodelet" type="nodelet" name="cloud_1_filter_fov_depth" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_1/depth_registered/points_rotated"/>
        <remap from="~output" to="/camera_1/depth_registered/points_filtered_depth"/>
        <rosparam>
            filter_field_name: y
            filter_limit_min: -0.15
            filter_limit_max:  0.15
            filter_limit_negative: False
        </rosparam>
    </node>	
    <node pkg="nodelet" type="nodelet" name="cloud_1_filter_fov_sides" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_1/depth_registered/points_filtered_depth"/>
        <remap from="~output" to="/camera_1/depth_registered/points_filtered_sides"/>
        <rosparam>
            filter_field_name: x
            filter_limit_min: -0.15
            filter_limit_max:  0.15
            filter_limit_negative: False
        </rosparam>
    </node>	
    <node pkg="nodelet" type="nodelet" name="cloud_1_ground_filter" args="load pcl/PassThrough pcl_manager" output="screen">
        <remap from="~input"  to="/camera_1/depth_registered/points_filtered_sides"/>
        <remap from="~output" to="/camera_1/depth_registered/points_filtered_ground_plane"/>
        <rosparam>
            filter_field_name: z
            filter_limit_min: 0.005
            filter_limit_max: 0.5
            filter_limit_negative: False
        </rosparam>
    </node>	


    <!-- Statistical outlier removal for good measure -->
    <node pkg="nodelet" type="nodelet" name="cloud_1_outlier_removal_pass1" args="load pcl/StatisticalOutlierRemoval pcl_manager" output="screen"> 
        <remap from="~input"  to="/camera_1/depth_registered/points_filtered_ground_plane" /> 
        <remap from="~output" to="/camera_1/depth_registered/points_filtered" /> 
        <rosparam> 
               mean_k: 100 
               stddev: 0.9 
               negative: False 
        </rosparam> 
    </node>

</launch>

